llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp

```diff
@@ -0,0 +1,504 @@
+}
+
+CASID GRPCRelayCAS::getID(ObjectRef Ref) const {
+  return getID(asInMemoryIndexValue(Ref));
+}
+
+CASID GRPCRelayCAS::getID(ObjectHandle Handle) const {
+  return getID(asInMemoryCASData(Handle).getIndex());
+}
+
+Optional<ObjectRef> GRPCRelayCAS::getReference(const CASID &ID) const {
+  assert(ID.getContext().getHashSchemaIdentifier() ==
+             getContext().getHashSchemaIdentifier() &&
+         "Expected ID from same hash schema");
+  auto &I = indexHash(ID.getHash());
+  return toReference(I);
+}
+
+ObjectRef GRPCRelayCAS::getReference(ObjectHandle Handle) const {
+  return toReference(asInMemoryCASData(Handle));
+}
+
+Expected<ObjectHandle> GRPCRelayCAS::load(ObjectRef Ref) {
+  auto &I = asInMemoryIndexValue(Ref);
+
+  // Return the existing value if we have one.
+  if (const InMemoryCASData *Existing = I.Data.load())
+    return getObjectHandle(*Existing);
+
+  // Send the request to remote to load the object. Since Generator function
+  // can't return nullptr to indicate failed load now, we can't lock the entry
+  // into busy state. we will send parallel loads and put one of the result into
+  // the local CAS.
+  auto &GetQueue = CASDB->getQueue();
+  GetQueue.getAsync(getDataIDFromRef(Ref));
+  auto Response = GetQueue.receiveNext();
+  if (!Response)
+    return Response.takeError();
+
+  SmallVector<const InMemoryIndexValueT *> InternalRefs;
+  for (auto &ID : Response->Refs) {
+    auto &Ref = indexHash(arrayRefFromStringRef(ID));
+    InternalRefs.push_back(&Ref);
+  }
+  if (!Response->BlobData)
+    return createStringError(std::make_error_code(std::errc::invalid_argument),
+                             "Expect result to be returned in buffer");
+
+  ArrayRef<char> Content(Response->BlobData->data(),
+                         Response->BlobData->size());
+  return getObjectHandle(storeObjectImpl(I, InternalRefs, Content));
+}
+
+uint64_t GRPCRelayCAS::getDataSize(ObjectHandle Handle) const {
+  return getInMemoryCASData(Handle).getDataSize();
+}
+
+Error GRPCRelayCAS::forEachRef(ObjectHandle Handle,
+                               function_ref<Error(ObjectRef)> Callback) const {
+  auto &Node = getInMemoryCASData(Handle);
+  for (const InMemoryIndexValueT *Ref : Node.getRefs())
+    if (Error E = Callback(toReference(*Ref)))
+      return E;
+  return Error::success();
+}
+
+ObjectRef GRPCRelayCAS::readRef(ObjectHandle Node, size_t I) const {
+  return toReference(*getInMemoryCASData(Node).getRefs()[I]);
+}
+
+size_t GRPCRelayCAS::getNumRefs(ObjectHandle Handle) const {
+  return getInMemoryCASData(Handle).getNumRefs();
+}
+
+ArrayRef<char> GRPCRelayCAS::getData(ObjectHandle Handle,
+                                     bool RequiresNullTerminator) const {
+  return getInMemoryCASData(Handle).getData();
+}
+
+Expected<ObjectRef>
+GRPCRelayCAS::storeFromOpenFileImpl(sys::fs::file_t FD,
+                                    Optional<sys::fs::file_status> FS) {
```

Comment:
This is not a new function. It always exists and used by CAS file system and it has a different default implementation overwrite in BuiltinCAS and here is another one for GRPC CAS.

BuiltinCAS use this to store/use the underlying memory mapped file region directly and GRPC uses the original file directly to send to server.

Reply:
Ok, I see it is supporting the `ObjectStore::storeFromOpenFile()` API.
It seems useful to add a `ObjectStore::storeFromFile(StringRef FilePath)`, so the caller doesn't go through hoops if they just have a file path (and the GRPC implementation just passes along the path), but this could be considered as follow-up.